lib/grppcaut.gi

```diff
@@ -869,12 +907,319 @@ end;
+  AddDictionary(d,pnt);
+  for p in orb do
+    for gen in acts do
+      i:=act(p,gen);
+      MakeImmutable(i);
+      if not KnowsDictionary(d,i) then
+	Add( orb, i );
+	if Length(orb)>lim then
+	  return true;
+	fi;
+	AddDictionary(d,i);
+      fi;
+    od;
+  od;
+  return false;
+end);
+
+# another function that should be more general and has space for improvement.
+BindGlobal("SubspaceStabilizerMatrixGroup",function(G,bas)
+local f,act,xset,stb,hom,n,m,l,i;
+  n:=DimensionOfMatrixGroup(G);
+  if Length(bas)=0 or ForAll(GeneratorsOfGroup(G),
+	     x->ForAll(bas,y->SolutionMat(bas,y*x)<>fail)) then
+    # space is fixed
+    return G;
+  fi;
+  bas:=OnSubspacesByCanonicalBasis(bas,One(G)); # canonical form
+  f:=DefaultFieldOfMatrixGroup(G);
+  if IsNaturalGL(G) then
+    # write down stabilizer
+    l:=Length(bas);
+    act:=BaseSteinitzVectors(IdentityMat(n,f),
+	  bas);
+    act:=Concatenation(bas,act.factorspace);
+    stb:=[];
+    # subspace
+    for i in GeneratorsOfGroup(GL(l,f)) do
+      m:=IdentityMat(n,f);
+      m{[1..l]}{[1..l]}:=i;
+      Add(stb,m);
+    od;
+    # factor
+    for i in GeneratorsOfGroup(GL(n-l,f)) do
+      m:=IdentityMat(n,f);
+      m{[l+1..n]}{[l+1..n]}:=i;
+      Add(stb,m);
+    od;
+    # mixer
+    m:=IdentityMat(n,f);
+    m[l+1][1]:=One(f);
+    Add(stb,m);
+    stb:=List(stb,x->x^act); # base change
+    stb:=Group(stb);
+    SetSize(stb,Size(GL(l,f))*Size(GL(n-l,f))*Size(f)^(l*(n-l)));
+    Info(InfoMatOrb,2,"reduced full GL by ",Size(G)/Size(stb));
+    return stb;
+  fi;
+  if not OrbitIsLonger(GeneratorsOfGroup(G),bas,OnSubspacesByCanonicalBasis,
+	   Size(f)^QuoInt(Length(bas[1]),2)) then
+    #just plain
+    act:=OnSubspacesByCanonicalBasis;
+    xset:=Orbit(G,bas,act);
+    stb:=[Position(xset,bas)];
+  else
+    act:=OnLines;
+    stb:=NormedVectors(VectorSpace(f,bas));
+    xset:=Union(Orbits(G,stb,act));
+    stb:=Set(List(stb,x->Position(xset,x)));
+  fi;
+  hom:=ActionHomomorphism(G,xset,act,"surjective");
+  act:=PreImage(hom,Stabilizer(Image(hom),stb,OnSets));
+  Info(InfoMatOrb,2,"reduce ",Size(G), " to ",Size(act));
+  return act;
+end);
+
+# construct subgroup of GL that stabilizes the spaces given and fixes teh
+# listed spaceorbits.
+
+# auxillary
+RedmatSpanningIndices:=function(gens)
+local bas,n,one,new,a,b,g;
+  n:=Length(gens[1]);
+  one:=One(gens[1]);
+  new:=[];
+  bas:=[];
+  while Length(bas)<n do
+    a:=First(one,x->Length(bas)=0 or SolutionMat(bas,x)=fail);
+    Add(new,Position(one,a));
+    Add(bas,a);
+    # spin
+    for b in bas do
+      for g in gens do
+        a:=b*g;
+	if SolutionMat(bas,a)=fail then
+	  Add(bas,a);
+	fi;
+      od;
+    od;
+  od;
+  return new;
+end;
+
+SpaceAndOrbitStabilizer:=function(n,field,spaces,osporb)
+local spaceincl,outvecs,l,sub,yet,i,j,k,s,t,new,incl,min,rans,sofar,done,
+      gens,one,spl,ngens,m,sz,a,sporb,notyet;
+
+  spaceincl:=function(big,small)
+    return RankMat(Concatenation(big,small))=Length(big);
+  end;
```

Comment:
Of course this becomes irrelevant if `spaceincl` is removed, as I suggest below :-)